Dockerfile

```diff
@@ -0,0 +1,116 @@
+# syntax=docker/dockerfile:1
+# This is loosely based on `docker init`'s rust template.
+# For a completely clean build, run something like this:
+# ```
+# docker build --build-arg=PROFILE=dev --build-arg=ENTRYPOINT=leader --no-cache
+# ```
+
+#############
+# Build stage
+#############
+# - `/src` is the repo directory.
+# - `/artifacts` is $CARGO_TARGET_DIR.
+# - `/output` is where the binaries go.
+
+ARG BUILD_BASE=rustlang/rust:nightly-bullseye-slim
+FROM ${BUILD_BASE} AS build
+
+# Install build dependencies.
+RUN apt-get update && apt-get install -y \
+    # for jemalloc
+    libjemalloc-dev \
+    libjemalloc2 \
+    make \
+    # for openssl
+    libssl-dev \
+    pkg-config \
+    # clean the image
+    && rm -rf /var/lib/apt/lists/*
+
+ARG PROFILE=release
+# forward the docker argument so that the script below can read it
+ENV PROFILE=${PROFILE}
+
+# Build the application.
+RUN \
+    # mount the repository so we don't have to COPY it in
+    --mount=type=bind,source=.,target=/src \
+    # cache artifacts and the cargo registry to speed up subsequent builds
+    --mount=type=cache,target=/artifacts \
+    --mount=type=cache,target=/usr/local/cargo/registry/ \
+    # run the build
+    <<EOF
+set -eux
+
+# need to change workdir instead of using --manifest-path because we need
+# .cargo/config.toml
+cd /src
+
+# use the cache mount
+# (we will not be able to to write to e.g `/src/target` because it is bind-mounted)
+CARGO_TARGET_DIR=/artifacts cargo build --locked "--profile=${PROFILE}" --all
+
+# narrow the find call to SUBDIR because if we just copy out all executables
+# we will break the cache invariant
+if [ "$PROFILE" = "dev" ]; then
+    SUBDIR=debug # edge case
+else
+    SUBDIR=$PROFILE
+fi
+
+# maxdepth because binaries are in the root
+# - other folders contain build scripts etc.
+mkdir /output
+find "/artifacts/$SUBDIR" \
```

Comment:
So we would find here not just `zero_bin` binaries, but any test or utility binary that exists in all crates in our workspace (and that we would potentially add in the future). Why would we need to use them in a containerized environment? 
Docker images that one project exports communicate to users what are the important deployment products of the project. It is like a contract to the DevOps observable in a few seconds - you should deploy this program/image to use/run our project.

Reply:
We still ship only two docker images - leader and worker. I don't think it's a blocker to _also_ ship e.g `verifier` in the docker image - do you have similar issues with shipping e.g debian's copy of `grep` in the image?

Reply:
Your objection has changed from
> Docker images that one project exports communicate 
(this PR doesn't change the docker images we publish) to 
> It is not just about what we ship, but with the average user sees when he opens the project docker files

Which you've already stated elsewhere[^1].
In addition, I'd argue that this `Dockerfile` is clearer than the previous ones - they're not hacking the workspace with giant sed commands[^2] just a simple `cargo build --all`

[^1]: https://github.com/0xPolygonZero/zk_evm/pull/357#discussion_r1663068377
[^2]: https://github.com/0xPolygonZero/zk_evm/blob/653b7e4ea42cbf77d6b723dd51a2bb41f14919eb/leader.Dockerfile#L9